### PR TITLE
Add conversion of arrays to std::vector<T>

### PR DESCRIPTION
**Pull request by joshessman-llnl**
# Summary

- This PR is a feature
- It does the following:
  - Adds a method to retrieve `std::vector`s of containers that ignores indices

Inlet currently uses `std::unordered_map<int, T>` as the primary representation for array types, as some `Reader`s support array discontiguous indexing.  In some cases only the contents of the array is useful (i.e., not its indices), so the added method provides a means of obtaining just the values.  Values are returned in ascending order by index.

This approach is more permissive than one that checks for contiguity and errors out accordingly, but since the user is specifically requesting a vector of the contents, I think it is reasonable to not require contiguity as long as the order of the returned vector is guaranteed.

### CODE REVIEW

File: src/axom/inlet/Table.hpp [hunk 1]
```diff
@@ -157,6 +157,24 @@ struct is_inlet_primitive_dict<std::unordered_map<VariantKey, T>>
   static constexpr bool value = is_inlet_primitive<T>::value;
 };
 
+template <typename T>
+struct is_std_vector : std::false_type
+{ };
+
+template <typename T>
+struct is_std_vector<std::vector<T>> : std::true_type
```
Comment (estebanpauli) [inline src/axom/inlet/Table.hpp:165]: In addition to the type parameter, std::vector has the allocator parameter. Given that Axom is going to be constructing the vector, its probably fine that it is restricted to the default allocator. It would probably be good to note that here, though. It might also be worth just adding the allocator type as a template parameter.

File: src/axom/inlet/Table.hpp [hunk 2]
```diff
@@ -764,6 +782,51 @@ class Table : public Verifiable<Table>
...
+   * 
+   * \return The values in the retrieved container
+   * 
+   * \tparam T The container type, i.e., T = std::vector<V>
+   * 
+   * \note Elements in the returned array will be in ascending order by index,
+   * regardless of index contiguity or base index
+   *******************************************************************************
+   */
+  template <typename T>
```
Comment (estebanpauli) [inline src/axom/inlet/Table.hpp:799]: See previous comment about the allocator parameter for vectors.

File: src/axom/inlet/tests/inlet_object.cpp [hunk 3]
```diff
@@ -436,6 +436,40 @@ TYPED_TEST(inlet_object, implicit_conversion_primitives)
...
+  Inlet inlet = createBasicInlet<TypeParam>(&ds, testString);
+
+  // Define schema
+  inlet.addIntArray("arr");
+
+  // Attempt both construction and assignment
+  std::vector<int> expected_arr {4, 6, 10};
+  std::vector<int> arr = inlet["arr"];
+  EXPECT_EQ(arr, expected_arr);
+  arr = inlet["arr"];
```
Comment (estebanpauli) [inline src/axom/inlet/tests/inlet_object.cpp:452]: Why is this here a second time? Does this test a separate path than the line above?

File: src/axom/inlet/Table.hpp (hunk 1 above)
Comment (kennyweiss) [reply src/axom/inlet/Table.hpp:165]: I don't think we'd want to complicate the interface in that way.

I was actually going to suggest that, in the future, this move from `std::vector` to `axom::Array`, since we're trying to eliminate usage of `std::vector` from our codebase.  Note that axom::Array uses umpire allocations instead of std::allocators.

Tag: @bmhan12

File: src/axom/inlet/tests/inlet_object.cpp (hunk 3 above)
Comment (kennyweiss) [reply src/axom/inlet/tests/inlet_object.cpp:452]: @estebanpauli -- Despite the syntax, Line 450 actually uses a copy constructor, rather than the assignment operator.

File: src/axom/inlet/tests/inlet_object.cpp [hunk 4]
```diff
@@ -639,6 +673,74 @@ TEST(inlet_object_lua, array_from_bracket)
...
+    "        { bar = false; baz = true} }";
+  DataStore ds;
+  Inlet inlet = createBasicInlet<axom::inlet::LuaReader>(&ds, testString);
+
+  auto& arr_table = inlet.addGenericArray("foo");
+
+  arr_table.addBool("bar", "bar's description");
+  arr_table.addBool("baz", "baz's description");
+  std::vector<Foo> expected_foos = {{true, false}, {false, true}};
+  auto foos = inlet["foo"].get<std::vector<Foo>>();
```
Comment (kennyweiss) [inline src/axom/inlet/tests/inlet_object.cpp:740]: This is really nice!

File: src/axom/inlet/tests/inlet_object.cpp [hunk 5]
```diff
@@ -639,6 +673,74 @@ TEST(inlet_object_lua, array_from_bracket)
...
+
+  // Define schema
+  inlet.addIntArray("arr");
+
+  // Attempt both construction and assignment
+  std::vector<int> expected_arr {4, 6, 10};
+  std::vector<int> arr = inlet["arr"];
+  EXPECT_EQ(arr, expected_arr);
+  arr = inlet["arr"];
+  EXPECT_EQ(arr, expected_arr);
```
Comment (kennyweiss) [inline src/axom/inlet/tests/inlet_object.cpp:690]: Minor: For at least one of these tests, it might be good to show the other way of accessing these using a map. Or perhaps to add a comment that the array getter removes the indices.

**Review by kennyweiss — APPROVED** [review]
Thanks @joshessman-llnl -- this is a nice contribution!

I added a note below about the possibility of changing this down the road from `std::vector` to `axom::Array`, which mostly be a drop-in replacement for `std::vector`, with some differences related to memory allocation/management/semantics.

Note that `axom::Array` is not yet ready, so my comments are meant to be informative rather than actionable, and the changes will be pretty straightforward.

File: src/axom/inlet/tests/inlet_object.cpp (hunk 3 above)
Comment (joshessman-llnl) [reply src/axom/inlet/tests/inlet_object.cpp:452]: Copy construction vs. copy assignment is a slightly different overload resolution context that has caused problems with these implicit conversions in the past, so this test checks both to avoid regressions.

Comment (estebanpauli) [reply src/axom/inlet/tests/inlet_object.cpp:452]: Sorry, I meant a separate path for the conversion from the result of inlet["arr"] (a Proxy, I believe), to something that the vector can be constructed/assigned from. I guess it is important to test both though, since as you point out, they are not the same.

**Review by white238 — APPROVED** [review]

**Comment by white238** [comment]
Thanks @joshessman-llnl !
